### ISSUE TITLE
[MODULES-13] Ability limit imported packages

### ISSUE THREAD

**Issue by dmlloyd** [jboss-modules/jboss-modules#383]
@tdiesler opened **[MODULES-13](https://issues.redhat.com/browse/MODULES-13?redirect=false)** and commented

Consider this

Bundle-SymbolicName: jbosgi323-bundleA
Import-Package: org.jboss.test.osgi.container.classloader.support.a

Bundle-SymbolicName: jbosgi323-bundleB
Export-Package:
org.jboss.test.osgi.container.classloader.support.a,
org.jboss.test.osgi.container.classloader.support.b

So moduleA has a dependency on moduleB because A imports package 'a' which B exports.
However, loading a class from package 'b' through A is expected to fail, because A does not import nor contain package 'b'

Currently this usecase fails (i.e. the class from b can be loaded through A) because you cannot control the paths that a module imports.

---
No further details from [MODULES-13](https://issues.redhat.com/browse/MODULES-13?redirect=false)

**Comment by dmlloyd** [comment]
@tdiesler commented

Done

http://github.com/jbosgi/jboss-modules/commit/9e63a2c5f7ce0d00f630e10a728ff2e855516b2c

OSGiModuleClassLoader.findClass now only calls the super ModuleClassLoader.findClass for the paths it truly imports.
If it is not an import (i.e. a class from the local content) we make a separate call to ModuleClassLoader.loadClassLocal

**Comment by dmlloyd** [comment]
@tdiesler commented

Please pull from http://github.com/jbosgi/jboss-modules/tree/mod13

**Comment by dmlloyd** [comment]
@dmlloyd commented

You should be able to do this already using DependencySpec.exportExcludes/exportIncludes (yeah I know they're called export* but it's supposed to control the exports of the dependency and thus your imports).  Does this not work?

**Comment by dmlloyd** [comment]
@tdiesler commented

http://community.jboss.org/message/554796#554796

**Comment by dmlloyd** [comment]
**[John Bailey](https://issues.redhat.com/secure/ViewProfile.jspa?name=johnbailey)** commented

Please review the updates in this repo:  git://github.com/baileyje/jboss-modules.git

I have added support for filtering imports in dependencies.

**Comment by dmlloyd** [comment]
@dmlloyd commented

Merged

**Comment by dmlloyd** [comment]
@tdiesler commented

Thanks John.

Skimming over it, it seems to address the concept that I need. I'll change the OSGi layer and try it out as soon as I can work with master again, which is currently blocked by MODULES-2

The modules schema does not seem to be in sync with this change.

**Comment by dmlloyd** [comment]
@dmlloyd commented

The schema fix has been moved to a new issue: MODULES-14

**Comment by dmlloyd** [comment]
@dmlloyd commented

Released.
